jwt-angular/src/app/auth.service.ts

```diff
@@ -0,0 +1,135 @@
+import { Injectable } from '@angular/core';
+import { Observable, of } from 'rxjs';
+import { tap, delay } from 'rxjs/operators';
+// esModuleInterop in your tsconfig
+import moment from 'moment';
+
+import { ApiService } from './api.service';
+
+export interface AuthData {
+  username: string;
+  password: string;
+}
+
+export interface AuthResponse {
+  access: string;
+  refresh?: string;
+  access_expiry: number;
+  refresh_expiry?: number;
+  csrf_token: string;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthService {
+  // Local storage
+  private ls = window.localStorage;
+  // Store auth info in memory
+  // CSRF token
+  csrfToken: string;
+  // Access expiry
+  private accessExpiry: number;
+  // Refresh expiry
+  private refreshExpiry: number;
+
+  constructor(private apiService: ApiService) {
+    // Syncronize other tabs
+    window.addEventListener('storage', this.syncAuth);
+  }
+
+  // Sign in
+  obtainBoth(data: AuthData): Observable<AuthResponse> {
+    return this.apiService.post(`/token/both/`, data)
+      .pipe(tap(response => {
+        this.setAuth(response);
+        this.scheduleRefresh();
+        this.ls.setItem('signin', moment().format());
+      }));
+  }
+
+  refreshToken(): Observable<AuthResponse> {
+    return this.apiService.post(`/token/access/`, {})
+      .pipe(tap(response => {
+        console.log(response);
+        this.setAuth(response);
+        this.scheduleRefresh();
+      }));
+  }
+
+  // Sign out
+  deleteToken(): Observable<null> {
+    return this.apiService.post(`/token/delete/`, {})
+      .pipe(tap(response => {
+        this.purgeAuth();
+        this.ls.setItem('signout', moment().format());
+      }));
+  }
+
+  isAuthenticated(): boolean {
+    if (!this.hasAuthData()) {
+      return false;
+    }
+    const now = moment();
+    const accessExpiry = moment(this.accessExpiry);
+    return now.isBefore(accessExpiry);
+  }
+
+  canRefresh(): boolean {
+    if (!this.hasAuthData()) {
+      return false;
+    }
+    const now = moment();
+    const refreshExpiry = moment(this.refreshExpiry);
+    return now.isBefore(refreshExpiry);
+  }
```

Comment:
Yes, I am trying to refresh before we hit the `401`. I thought this could be a nice addition.

Reply:
I think the main difference with mobile is that you can have several sessions (or tabs) for the same user.
As I mentioned we already always perform a `refresh token` when the application is loaded.
This means that for every page load I would need to hit 2 `401` responses before I can be sure that my user isn't authenticated. I guess I was trying to avoid this.

- https://hasura.io/blog/best-practices-of-using-jwt-with-graphql/ (similar approach)

Reply:
> I think the main difference with mobile is that you can have several sessions (or tabs) for the same user.

Not really a difference but more of a similarity. It's the same as a web browser in that you're saving the tokens once. I guess the only difference is that we also save their username + password in a keychain so we can always log the user in whenever he/she opens the app (persistence).

If you get a 401, you refresh. You might be thinking "several sessions" with multiple devices; you must remember JWT is state _less_ authentication.

> This means that for every page load I would need to hit 2 401 responses before I can be sure that my user isn't authenticated.

Sorry I didn't catch that from the beginning, but that's good that it's a 2 hit 401 response mechanism :)

Don't you think that refreshing the token every new page is a little inefficient? It seems much better to have that scheduled refresh expand to be 10 seconds (accounting for poor network conn) before expiration rather than one second so you can avoid two HTTP calls, which starts to amount costs quite a bit for people hosting on EC2 (data transfer out) or Heroku.

The two things that an application would have would be:

1. the first 401 response would require a token refresh. If that returns 401, then logout
2. scheduled refreshing of the token.

It's a lot of wasted network resources (and honestly not needed) to continuously call `refresh-token` every new page, imo.

Reply:
> Don't you think that refreshing the token every new page is a little inefficient?

The problem is that there is no way to tell if the user is authenticated when you are in a new page and bear in mind that we have to add the CSRF token in the header of every HTTP request. CSRF token we don't have access to because it is in a `httpOnly` cookie. So unless you want me to store the CSRF token in `localStorage`, there is really no way I would be able to do this differently with the current setup.

Now I think that I could try to implement the 2 hits mechanism in the interceptor. It isn't the case at the moment. The code will be a little tricky to understand for Angular and rxjs beginners but I could totally implement such an approach.
The question is then do we use

- scheduled refresh
- 2 hits 401
- or a combination of both

> It seems much better to have that scheduled refresh expand to be 10 second.

yes but then we need to have access tokens that are valid for more than 10 seconds. :laughing: The default one was even shorter, 5 seconds! I thought you wanted to keep it small for demonstration purposes.

Reply:
I'd still prefer the 2 hits 401 method. That way, you won't need to store the CSRF token in the local storage, right?

> yes but then we need to have access tokens that are valid for more than 10 seconds. 😆 The default one was even shorter, 5 seconds! I thought you wanted to keep it small for demonstration purposes.

Haha true. The reason I kept it at 5 seconds was because I used the "2 hits 401" method for my mobile application. Using scheduled refreshes would've consumed too much computational resources for a simple background timer on both Android and iOS.

Extending it to 10 seconds would be practical for those in poor network connections for scheduled refresh method. I'd recommend changing it to 10 seconds, BUUUUT:

I believe the 2 hits 401 is the best method and should be the only method. Take a look at my implementation for iOS:

https://github.com/Andrew-Chen-Wang/mobile-auth-example/blob/c51df854d8fa8bf9e15e5f162507aeed32bde998/jwt-ios/jwt-ios/Networking/Service/Router.swift#L27-L48

The rundown is as follows: if a random endpoint, for example "Ping" returned a 401 error, since we already know the user must be logged in to have even gotten "visually seen" the Ping endpoint, refresh the token. You can see the refreshing process in L29 in which I immediately refresh the token. If it's a success, then I retry the initial request; if it failed, I would have to use the saved username + password to get both tokens again.

For this repository, however, you won't be getting to that second portion (i.e. the "getting both tokens again" part). Instead, that would be a simple log out of the current user in the browser. This is because the developer shouldn't be saving the user's password in cookies. Thus I imparted the MITM attack in a different comment. If a thief took the refresh token, due to its stateless nature, it should expire, log out the thief, and the least amount damage could've been done.

In other words, I suggest the 2 hits 401. It's not bad to have the scheduled refresh thing, and you could possibly still include it in perhaps a second directory demonstrating both ways (i.e. scheduled refresh and 2 hits 401), but keep in mind the costs of constantly having to refresh a token, especially if a user, especially someone like my dad, just leaves the browser open forever. Unnecessary traffic; unnecessary costs, computationally and economically.

> The code will be a little tricky to understand for Angular and rxjs beginners

When I was looking back at my Android code, I couldn't understand half of it; too complex. That's why I showed the Swift one lol. That's also why I liked the idea of setting up template repositories (even for myself with the [mobile-auth-example](https://github.com/Andrew-Chen-Wang/mobile-auth-example); you don't need to understand it, just use it and you'll be _mostly_ fine.

Reply:
> I'd still prefer the 2 hits 401 method. That way, you won't need to store the CSRF token in the local storage, right?

No. 2 reasons:
- all unsecure requests will fail because you cannot insert the CSRF token in the HTTP header, you don't have access to it with javascript because it's in a httpOnly cookie.
- the ping would work because of the cookies, and I guess you could set a boolean to tell the user is authenticated, but it wouldn't return the CSRF token. So you would still need to get it in some way. The only way you can share data between pages in a browser is via a Cookie not http only or localStorage.

> keep in mind the costs of constantly having to refresh a token, especially if a user, especially someone like my dad, just leaves the browser open forever.

Yes but there is a very simple way to tell if page/tab is active (visibility API) or not. So we could easily block the referesh when the page is not active and resschedule when the user comes back to it.

I think that you are right in general, we need the 2 hits, just because it would mitigate poor connectivity issues, but it's not a golden ticket either.

In short scheduled, in good conditions would reduce round trips to the server and 2 hits would increase robustness of the system.

Reply:
Note that in you have a webapp where users have multiple tabs open all the time, scheduling becomes inefficient.